### PR TITLE
SLVS-1288 Provide the rules configuration when initializing SlCore

### PR DESCRIPTION
**Pull request by gabriela-trutan-sonarsource**

### CODE REVIEW

File: src/Core/UserRuleSettings/UserSettingsConstants.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.IO;
+
+namespace SonarLint.VisualStudio.Core.UserRuleSettings;
+
+public static class UserSettingsConstants
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Core/UserRuleSettings/UserSettingsConstants.cs:25]: This should be a property of IUserSettingsProvider

File: src/SLCore/Analysis/SLCoreRuleSettings.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+using SonarLint.VisualStudio.Core.UserRuleSettings;
+using SonarLint.VisualStudio.SLCore.Core;
+using SonarLint.VisualStudio.SLCore.Service.Rules;
+using SonarLint.VisualStudio.SLCore.Service.Rules.Models;
+
+namespace SonarLint.VisualStudio.SLCore.Analysis;
+
+public interface ISLCoreRuleSettings
+{
+    Dictionary<string, StandaloneRuleConfigDto> RulesSettings { get; }
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore/Analysis/SLCoreRuleSettings.cs:32]: Convert to Get.. method as this property does some calculation inside and does not cache the value

File: src/SLCore/Analysis/SLCoreRuleSettings.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    [ImportingConstructor]
+    public SlCoreRuleSettings(ILogger logger, IUserSettingsProvider userSettingsProvider, ISLCoreServiceProvider slCoreServiceProvider)
+    {
+        this.logger = logger;
+        this.userSettingsProvider = userSettingsProvider;
+        this.slCoreServiceProvider = slCoreServiceProvider;
+    }
+
+    public Dictionary<string, StandaloneRuleConfigDto> RulesSettings => MapRuleSettingsToSlCoreSettings(userSettingsProvider.UserSettings.RulesSettings);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore/Analysis/SLCoreRuleSettings.cs:52]: public -> private -> public -> private is hard to read

File: src/SLCore/Analysis/SLCoreRuleSettings.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+using SonarLint.VisualStudio.SLCore.Core;
+using SonarLint.VisualStudio.SLCore.Service.Rules;
+using SonarLint.VisualStudio.SLCore.Service.Rules.Models;
+
+namespace SonarLint.VisualStudio.SLCore.Analysis;
+
+public interface ISLCoreRuleSettings
+{
+    Dictionary<string, StandaloneRuleConfigDto> RulesSettings { get; }
+    void UpdateStandaloneRulesConfiguration();
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore/Analysis/SLCoreRuleSettings.cs:33]: It feels like these 2 things don't really belong together. We have UserSettingsUpdater and UserSettingsProvider split, but SLCoreSettings provider and updater are in the same class together

File: src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs [hunk 5]
```diff
@@ -477,7 +483,42 @@ private void SetUpThreadHandling(IThreadHandling threadHandling)
             compatibleNodeLocator,
             activeSolutionBoundTracker,
             configScopeUpdater,
-            threadHandling);
+            threadHandling, 
+            Substitute.For<ISLCoreRuleSettings>());
+    }
+
+    private SLCoreInstanceHandle CreateTestSubject(out ISLCoreRpcFactory slCoreRpcFactory,
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs:490]: this feels like a copy-paste just to add 1 parameter, why?

File: src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs [hunk 6]
```diff
@@ -405,6 +410,7 @@ public void Dispose_NotInitialized_DoesNothing()
         });
         jarLocator.ListJarFiles().Returns(JarList);
         activeSolutionBoundTracker.CurrentConfiguration.Returns(new BindingConfiguration(Binding, SonarLintMode.Connected, "dir"));
+        slCoreRuleSettings?.RulesSettings.Returns(new Dictionary<string, StandaloneRuleConfigDto>());
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs:413]: No unit tests where you actually verify that this value is ever used anywhere

File: src/SLCore.IntegrationTests/SLCoreTestRunner.cs [hunk 7]
```diff
@@ -148,12 +157,17 @@ private void SetUpLocalFolders()
...
 
         Directory.CreateDirectory(privateFolder);
         Directory.CreateDirectory(storageRoot);
         Directory.CreateDirectory(workDir);
         Directory.CreateDirectory(userHome);
     }
+
+    private static void DeleteDirectoryForLongPaths(string dirToDelete)
+    {
+        Directory.Delete(@"\\?\" + dirToDelete, true);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.IntegrationTests/SLCoreTestRunner.cs:171]: Ideally should be a separate PR as it is an unrelated change. And how does it work?

File: src/SLCore.IntegrationTests/FileAnalysisTests.cs [hunk 8]
```diff
@@ -109,25 +110,77 @@ public async Task StandaloneRuleConfig_CtorParamsAboveThreshold_JavaScriptActive
...
+        var issuesByFileUri = await RunFileAnalysis(ThreeSecretsIssuesPath, initialRulesConfig:secretsRuleConfig, updatedRulesConfig:null);
+
+        issuesByFileUri.Should().HaveCount(1);
+        issuesByFileUri[new FileUri(GetFullPath(ThreeSecretsIssuesPath))].Should().HaveCount(1);
+    }
+
     private async Task<Dictionary<FileUri, List<RaisedIssueDto>>> RunFileAnalysis(string fileToAnalyzeRelativePath, Dictionary<string, StandaloneRuleConfigDto> ruleConfigByKey = null)
     {
-        const string configScopeId = "ConfigScope1";
+        return await RunFileAnalysis(fileToAnalyzeRelativePath, null, ruleConfigByKey);
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore.IntegrationTests/FileAnalysisTests.cs:137]: why overload and not optional parameters?

File: src/Integration.Vsix/Analysis/UserSettingsUpdater.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+{
+    /// <summary>
+    /// Notification that one or more settings have changed
+    /// </summary>
+    event EventHandler SettingsChanged;
+
+}
+
+[Export(typeof(IUserSettingsUpdater))]
+internal sealed class UserSettingsUpdater : IUserSettingsUpdater, IDisposable
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix/Analysis/UserSettingsUpdater.cs:38]: This refactoring should have been a separate PR before the SLCore integration

File: src/Integration.Vsix/Analysis/UserSettingsUpdater.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+ */
+
+using System.ComponentModel.Composition;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.UserRuleSettings;
+using SonarLint.VisualStudio.SLCore.Analysis;
+
+namespace SonarLint.VisualStudio.Integration.Vsix.Analysis;
+
+public interface IUserSettingsUpdater
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration.Vsix/Analysis/UserSettingsUpdater.cs:28]: The name is kind of misleading. It does the updates but it's also a source of events about updates. I get why it's a separate class, so it's just a naming question

File: src/SLCore/Analysis/SLCoreRuleSettings.cs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+    public Dictionary<string, StandaloneRuleConfigDto> RulesSettings => MapRuleSettingsToSlCoreSettings(userSettingsProvider.UserSettings.RulesSettings);
+
+    private Dictionary<string, StandaloneRuleConfigDto> MapRuleSettingsToSlCoreSettings(RulesSettings rulesSettings)
+    { 
+        return rulesSettings.Rules.ToDictionary(kvp => kvp.Key, kvp => MapStandaloneRuleConfigDto(kvp.Value));
+    }
+
+    public void UpdateStandaloneRulesConfiguration()
+    {
```
Comment (georgii-borovinskikh-sonarsource) [inline src/SLCore/Analysis/SLCoreRuleSettings.cs:60]: This method is not expected to be called on the UI thread, right?

**Review by georgii-borovinskikh-sonarsource — CHANGES_REQUESTED** [review]

File: src/Core/UserRuleSettings/UserSettingsConstants.cs (hunk 1 above)
Comment (gabriela-trutan-sonarsource) [reply src/Core/UserRuleSettings/UserSettingsConstants.cs:25]: It can not be a property, because then I have to provide it in the UserSettingUpdater constructor to the SingleFileMonitor constructor creation. This makes the importing MEF constructor test to fail (because the property is null at the point when the constructor is called) So it has to be a static, so I thought a separate class makes more sense.

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 3 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:52]: Can you be more explicit what are you referring to, please?

File: src/Integration.Vsix/Analysis/UserSettingsUpdater.cs (hunk 10 above)
Comment (gabriela-trutan-sonarsource) [reply src/Integration.Vsix/Analysis/UserSettingsUpdater.cs:28]: Any proposal for a better name?

File: src/Integration.Vsix/Analysis/UserSettingsUpdater.cs (hunk 9 above)
Comment (gabriela-trutan-sonarsource) [reply src/Integration.Vsix/Analysis/UserSettingsUpdater.cs:38]: Sorry. I am not really used with PRs. I shall try to keep in mind next time.

File: src/SLCore.IntegrationTests/FileAnalysisTests.cs (hunk 8 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.IntegrationTests/FileAnalysisTests.cs:137]: Because they are more speaking. Optional parameters some disadvantages, like the fact  the number of optional parameters tend to increase (I have seen that a lot of unit tests methods have > 7 optional parameter), which is makes code very ugly and difficult to maintain.

File: src/SLCore.IntegrationTests/SLCoreTestRunner.cs (hunk 7 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.IntegrationTests/SLCoreTestRunner.cs:171]: It is related to the integration test, that was failing for me locally. 

The @"\\?\" prefix is a special syntax used to handle long paths in Windows. Normally, Windows paths are limited to 260 characters, but using the @"\\?\" prefix allows you to work with paths up to 32,767 characters.

https://learn.microsoft.com/en-gb/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 4 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:33]: Yes, because the UserSettingsUpdater, which is in the Integration.Vsix assembly, is responsible to update the all the RuleSettings. 

UserSettingsProvider, which is in the Core assembly takes care of the settings.json file.

Finally, the SlCoreSettings takes care of sending the settings to SlCore. So I think it makes sense that the settings and sending the update to SlCore should be together. Would be more confusing for the consumers if I split the logic into multiple classes.

File: src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs (hunk 6 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs:413]: Good point.

File: src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs (hunk 5 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs:490]: Otherwise, I would have to update 9 unit tests just because one more parameter was added, when those unit tests do not care at all about the rules settings. That's why this coding style with out parameters is so ugly.

But yes, doing that leads to duplicate code, so I have to find a better way.

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 11 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:60]: Right

File: src/Core/UserRuleSettings/UserSettingsConstants.cs (hunk 1 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/Core/UserRuleSettings/UserSettingsConstants.cs:25]: Create export supports providing mocks with custom behavior. Ideally, the file watcher should be wrapped with a factory interface and mocked, so that we don't actually create the file watcher in tests.

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 2 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:32]: To me it's just a bit of a misuse of the property. It's a fetch + convert, which is a complicated operation, and that should be a method. https://stackoverflow.com/questions/1374273/when-to-use-properties-instead-of-functions#:~:text=Use%20a%20property%20when%20the,ToString.

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 3 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:52]: The method ordering. You have a public method, followed by a private method, followed by a public method and then a private method again

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 4 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:33]: For example, `SLCoreInstanceHandle` would never need to access `UpdateStandaloneRulesConfiguration`, so, technically, the interface is not minimal. To me it still feels like a provider and a updater got merged together.

File: src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs (hunk 5 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs:490]: Having this out-parameter pattern is not ideal, so if we have more time I would refactor this, but as it is now, it's better to just add the new parameter to the existing method and update the usages.

File: src/SLCore.IntegrationTests/SLCoreTestRunner.cs (hunk 7 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore.IntegrationTests/SLCoreTestRunner.cs:171]: Have you tried this? https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=powershell#registry-setting-to-enable-long-paths

File: src/SLCore.IntegrationTests/FileAnalysisTests.cs (hunk 8 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore.IntegrationTests/FileAnalysisTests.cs:137]: But they're not more speaking if you use the same name for them. Rename the first one to `RunFileAnalysisWithDefaultRuleConfiguration`.

File: src/Integration.Vsix/Analysis/UserSettingsUpdater.cs (hunk 10 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/Integration.Vsix/Analysis/UserSettingsUpdater.cs:28]: 🤷

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 11 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:60]: What do you think about adding ThrowIfOnUIThread call here? I'm not sure if we should

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 3 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:52]: We should talk about in the weekly

File: src/Core/UserRuleSettings/UserSettingsConstants.cs (hunk 1 above)
Comment (gabriela-trutan-sonarsource) [reply src/Core/UserRuleSettings/UserSettingsConstants.cs:25]: Will be done as part of the architecture refactoring, in https://sonarsource.atlassian.net/browse/SLVS-1301

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 3 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:52]: Will be done as part of the architecture refactoring, in https://sonarsource.atlassian.net/browse/SLVS-1301

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 4 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:33]: Will be done as part of the architecture refactoring, in https://sonarsource.atlassian.net/browse/SLVS-1301

File: src/SLCore/Analysis/SLCoreRuleSettings.cs (hunk 11 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore/Analysis/SLCoreRuleSettings.cs:60]: Added it to the weekly discussion

File: src/SLCore.IntegrationTests/FileAnalysisTests.cs (hunk 8 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.IntegrationTests/FileAnalysisTests.cs:137]: Done, but the renaming is a little bit different.

File: src/SLCore.IntegrationTests/SLCoreTestRunner.cs (hunk 7 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.IntegrationTests/SLCoreTestRunner.cs:171]: Yes. Locally I could delete the folders, but the unit tests were failing

File: src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs (hunk 5 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs:490]: Simplified the discussion, by removing both overloads and using TestInitialize. It was a quick refactoring.

File: src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs (hunk 6 above)
Comment (gabriela-trutan-sonarsource) [reply src/SLCore.UnitTests/SLCoreInstanceHandleTests.cs:413]: Done.

**Review by georgii-borovinskikh-sonarsource — APPROVED** [review]
